docs/guides/versioning.md

```diff
@@ -0,0 +1,312 @@
+    padding-top: $spacing-02;
+    width: 100%;
+  }
+}
+```
+
+```diff
+function CardTitle({ children, title }) {
+  return (
+-   <span className="card--title" title={title}>
++   <span className={`${iotPrefix}--card--title`} title={title}>
+      {children}
+    </span>
+  );
+}
+```
+
+In general, we consider selectors to be internals/implementation details for the following reasons:
+
+1. There is no way we could test to see if a change we make could break how someone is overriding a specific selector
+1. Following semantic versioning with selectors would introduce broad implications resulting in the inflexibility or freezing of all aspects of our styles which would make it challenging to update without considering it a breaking change.
+
+We realize this can be painful for consuming projects. In attempt to ease this pain, here are a few supporting thoughts you may find helpful:
+
+1. Overriding styles from this library in your application is a valid use case that we would like to assist in supporting. While there is no way for this library to test your codebase for overriden styles, [version release changelogs](https://github.com/IBM/carbon-addons-iot-react/releases) are provided for each release, which contain a list of included commits and recently closed issues. This can be used to identify high level potential changes when upgrading versions of this library in your application. Additionally you may consider writing automated tests against styles you override to ensure their validity when updating.
+1. Changes to the public component API is tracked via [a jest snapshot file](https://github.com/IBM/carbon-addons-iot-react/blob/master/src/utils/__tests__/__snapshots__/publicAPI.test.js.snap). This includes `className` and other prop changes. The file can be used during your upgrade process to compare changes between different versions of the library. [Here is an example comparing v2.55.0 -> v2.56.0](https://github.com/IBM/carbon-addons-iot-react/compare/v2.55.0...v2.56.0#diff-789e184fd0335570057a913379ee5f43), tags can be changed and selected via the dropdowns at the top of that page.
+1. For testing automation, we are working to provide a dedicated `data-testid` data attribute placed on the containing element of each component. This should be used in place of class or id selectors in your testing automation. More info is available in the [testing guide documentation](/testing).
+
+#### A `data-testid` attribute is added
+
+semver bump: **minor**
+
+```diff
+- function ExportedComponent({ message }) {
++ function ExportedComponent({ testID, message }) {
+  return (
+    <>
+-     <span>{message}</span>
++     <span data-testid={testID}>{message}</span>
+    </>
+  );
+}
+
+ExportedComponent.propTypes = {
+  message: PropTypes.string,
++ testID: PropTypes.string,
+};
+ExportedComponent.defaultProps = {
+  message: '',
++ testID: 'ExportedComponent',
+};
+```
+
+#### A `data-testid` attribute default prop value is changed or removed
```

Comment:
While not clearly specified here, moving a data-test-id to some other element should be considered a breaking change.

Reply:
@stuckless Similar to IDs on DOM nodes, would it be more helpful if we could provide release notes listing these changes? A potential stretch goal could be to include codemods/scripts to assist in automating the migration.

Reply:
This is where semantic versioning works well for a library.  I know that if move to version 1.1 to version 1.2 or 1.1.1 there are no breaking changes.  My code will run just fine, and you just added new features or fixed a bug.   But in a UI component library you would need to knwo the same thing, and if you are introducing changes that will break testing, then that is a breaking change.  If as a consumer I have to assume that 1.1.1 and 2.0 needs the same amount of regression since while you having broken any APIs you may have broken testing.  So in planning an upgrade you always have to plan for the worse, since the semantic versioning means very little if you can't trust it.  One of the purposes of semantic versioning is to help a downstream consumer identify the risk involved in upgrading to a newer version of the library.

Reply:
👍 I've added a new section for `The DOM node that an `data-testid` corresponds to is changed` describing this as a breaking change.